### PR TITLE
[feat]: [IAC-45]: Add IACM Manager to main mono repo

### PR DESCRIPTION
**Pull request by nassergonzalez**
## Describe your changes

This changes are going to introduce the skeleton for the IACM manager. This manager is going to be required to interact with the PMS SDK and the new IACM module
This change also modifies the IACM FF name

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)

### CODE REVIEW

File: .aeriform/BUILD.bazel [hunk 1]
```diff
@@ -79,6 +79,9 @@ aeriform("//295-cdng-contracts:module")
 aeriform("//300-cv-nextgen:module")
 
 aeriform("//300-cv-nextgen:tests")
+aeriform("//310-iacm-manager:module")
+aeriform("//311-iacm-utilities:module")
+aeriform("//312-iacm-beans:module")
```
Comment (r0610205) [inline .aeriform/BUILD.bazel:84]: There are no beans anymore, right?

File: 332-ci-manager/config/ci-manager-config.yml [hunk 2]
```diff
@@ -243,7 +243,7 @@ pmsGrpcClientConfig:
 
 apiUrl: https://localhost:8181/#/
 
-shouldConfigureWithPMS: ${SHOULD_CONFIGURE_WITH_PMS:-false}
```
Comment (r0610205) [inline 332-ci-manager/config/ci-manager-config.yml:246]: Looks like the change from local env

Comment (nassergonzalez) [reply 332-ci-manager/config/ci-manager-config.yml:246]: you are right, sorry :(

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 14 of 65 files at r1, 2 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)* status: 49 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/filter/IACMStageFilterJsonCreator.java` line 122 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrW1d66rzCQq-pnYr:-NGyrW1d66rzCQq-pnYs:b-5zs24s) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/filter/IACMStageFilterJsonCreator.java#L122)):*
> ```Java
> 
>     log.info("Successfully created filter for integration stage {}", stageNode.getIdentifier());
>     return ciFilterBuilder.build();
> ```

Do we need these filters in pipeline list? plus why many validation are there here and not done via schema?



_Code quote:_
```Java
    log.info("Received filter creation request for integration stage {}", stageNode.getIdentifier());
    String accountId = filterCreationContext.getSetupMetadata().getAccountId();
    String orgIdentifier = filterCreationContext.getSetupMetadata().getOrgId();
    String projectIdentifier = filterCreationContext.getSetupMetadata().getProjectId();

    BaseNGAccess baseNGAccess = BaseNGAccess.builder()
                                    .accountIdentifier(accountId)
                                    .orgIdentifier(orgIdentifier)
                                    .projectIdentifier(projectIdentifier)
                                    .build();
    YamlField variablesField =
        filterCreationContext.getCurrentField().getNode().getField(YAMLFieldNameConstants.VARIABLES);
    if (variablesField != null) {
      FilterCreatorHelper.checkIfVariableNamesAreValid(variablesField);
    }

    CIFilterBuilder ciFilterBuilder = CIFilter.builder();
    CodeBase ciCodeBase = null;
    try {
      YamlNode properties =
          YamlUtils.getGivenYamlNodeFromParentPath(filterCreationContext.getCurrentField().getNode(), PROPERTIES);
      YamlNode ciCodeBaseNode = properties.getField(CI).getNode().getField(CI_CODE_BASE).getNode();
      ciCodeBase = IntegrationStageUtils.getCiCodeBase(ciCodeBaseNode);
    } catch (Exception ex) {
      // Ignore exception because code base is not mandatory in case git clone is false
      log.warn("Failed to retrieve ciCodeBase from pipeline");
    }

    if (ciCodeBase != null) {
      if (ciCodeBase.getRepoName().getValue() != null) {
        ciFilterBuilder.repoName(ciCodeBase.getRepoName().getValue());
      } else if (ciCodeBase.getConnectorRef().getValue() != null) {
        try {
          ConnectorDetails connectorDetails =
              connectorUtils.getConnectorDetails(baseNGAccess, ciCodeBase.getConnectorRef().getValue());
          String repoName = getGitRepo(connectorUtils.retrieveURL(connectorDetails));
          ciFilterBuilder.repoName(repoName);
        } catch (Exception exception) {
          log.warn("Failed to retrieve repo");
        }
      }
    }

    validateStage(stageNode);

    log.info("Successfully created filter for integration stage {}", stageNode.getIdentifier());
    return ciFilterBuilder.build();
```
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java` line 114 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrhIL7njJzVXZRq_8:-NGyrhIL7njJzVXZRq_9:bx6yj5u) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java#L114)):*
> ```Java
>     // on 3 elements, cloneCodebase, Infastructure (to use the dlite delegates) and Execution. I think that if any of
>     // the other values are present we should fail the execution
>     IntegrationStageConfig integrationStageConfig = (IntegrationStageConfig) stageNode.getStageInfoConfig();
> ```

Why your stage is same as CI? What will be your stage type?
___
*[`970-ng-commons/src/main/java/io/harness/ModuleType.java` line 28 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrway3_MH0Kg2kOHs:-NGyrway3_MH0Kg2kOHt:b-571pva) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/970-ng-commons/src/main/java/io/harness/ModuleType.java#L28)):*
> ```Java
>   @JsonProperty("STO") STO("Security Testing Orchestration"),
>   @JsonProperty("IACM")
>   IACM("Infrastructure as Code Manager", false,
> ```

Always add enum to the last

**Comment by nassergonzalez** [comment]
trigger smartchecks

**Comment by nassergonzalez** [comment]
@shubham149 IACM is the new module "Infrastructure as Code Manager" you can find all the documents that we are generating here: https://harness.atlassian.net/wiki/spaces/CI/pages/21177139903/IaCM+Architecture
and if you want to check the product specs you can find them here: https://harness.atlassian.net/wiki/spaces/CI/pages/21129920883/Infrastructure+as+Code+-+Product+Spec

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/registrars/IACMExecutionRegistrar.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import io.harness.ci.states.codebase.CodeBaseStep;
+import io.harness.ci.states.codebase.CodeBaseTaskStep;
+import io.harness.pms.contracts.steps.StepType;
+import io.harness.pms.sdk.core.steps.Step;
+import io.harness.registrar.NGCommonUtilStepsRegistrar;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class IACMExecutionRegistrar {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/registrars/IACMExecutionRegistrar.java:24]: This should be part IACM module

File: 332-ci-manager/service/src/main/java/io/harness/ci/serializer/CiBeansRegistrars.java [hunk 4]
```diff
@@ -51,6 +51,7 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import java.util.Arrays;
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/serializer/CiBeansRegistrars.java:54]: This is not required

File: 332-ci-manager/service/src/main/java/io/harness/ci/beans/steps/StepSpecTypeConstants.java [hunk 5]
```diff
@@ -36,6 +36,6 @@ public interface StepSpecTypeConstants {
 
   String CI_STAGE = "CI";
   String SECURITY_STAGE = "SecurityTests";
-
+  String IACM_STAGE = "IACM";
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/beans/steps/StepSpecTypeConstants.java:39]: Why is it residing in ci manager ?

File: 332-ci-manager/service/BUILD.bazel [hunk 6]
```diff
@@ -426,6 +426,7 @@ java_library(
         "//332-ci-manager/service/src/main/resources:resource",
     ],
     visibility = [
+        "//310-iacm-manager:__subpackages__",
```
Comment (harshjain12) [inline 332-ci-manager/service/BUILD.bazel:429]: I think it should be vice versa.
iacm manager should have dependency on ci manager service.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/registrars/IACMExecutionRegistrar.java (hunk 3 above)
Comment (nassergonzalez) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/registrars/IACMExecutionRegistrar.java:24]: Done

**Comment by shubham149** [comment]
@nassergonzalez I tried to look at the architecture docs but I was unable to figure out answer for "What does infrastructure as code" means? Can you explain it in simple words? Is it similar to terraform? Also, does architecture doc require some pre requisites? Maybe adding summary and problem statement to the doc will help.

File: 332-ci-manager/service/src/main/java/io/harness/ci/beans/steps/StepSpecTypeConstants.java (hunk 5 above)
Comment (nassergonzalez) [reply 332-ci-manager/service/src/main/java/io/harness/ci/beans/steps/StepSpecTypeConstants.java:39]: Moved to our manager

File: 332-ci-manager/service/BUILD.bazel (hunk 6 above)
Comment (nassergonzalez) [reply 332-ci-manager/service/BUILD.bazel:429]: How can I do that, I am not good with Java or its dependency method and I was able to do that following what STO did but idk how to invert the dependency

**Comment by nassergonzalez** [comment]
@shubham149 Basically is a new module that is going to help customers deploy their infrastructure and be able to provide insight in their infrastructures, like cost changes, TTL, resources created/destroyed/ audit trail, OPA policies..etc.

If you want more information about the architecture, you can speak with @allofthesepeople (Roger Simms) or if you want more information about the product with Uri. I am also available if you want to chat with me (just ping me a message and we ca set a call)

File: 332-ci-manager/service/BUILD.bazel (hunk 6 above)
Comment (nassergonzalez) [reply 332-ci-manager/service/BUILD.bazel:429]: As mentioned in our chat:
There is a STO dependency there also:
   "//310-iacm-manager:__subpackages__",
        "//315-sto-manager:__subpackages__",
        
 And removing this part of the code gives this error:
 
 (11:08:07) ERROR: /Users/ngonzalez/IdeaProjects/portal2/harness-core/310-iacm-manager/BUILD.bazel:419:12: in java_binary rule //310-iacm-manager:module: target '//332-ci-manager/service:module' is not visible from target '//310-iacm-manager:module'. Check the visibility declaration of the former target if you think the dependency is legitimate


It looks that this line is just so my module is able to see the 332 module

**Review by harshjain12 — COMMENTED** [review]
> Reviewed 41 of 65 files at r1, 2 of 7 files at r2, 1 of 7 files at r4, 3 of 8 files at r7, 8 of 10 files at r8, 13 of 13 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)* status: 1 change requests. 0 of 1 approvals obtained.

**Review by harshjain12 — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)* status: :shipit: complete! ✓

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 10 files at r8, 1 of 13 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)* status: 17 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/filter/IACMStageFilterJsonCreator.java` line 122 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrW1d66rzCQq-pnYr:-NHxTv8_3AWq6S8EHuMN:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/filter/IACMStageFilterJsonCreator.java#L122)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Do we need these filters in pipeline list? plus why many validation are there here and not done via schema?
</blockquote></details>

done
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java` line 114 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrhIL7njJzVXZRq_8:-NHxTz3ADMN05DI9FOjH:b-16axlr) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java#L114)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Why your stage is same as CI? What will be your stage type?
</blockquote></details>

To be check with CI Team cc: @harshjain12 
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/step/IACMPMSStepPlanCreator.java` line 18 at r11](https://reviewable.io/reviews/harness/harness-core/39942#-NHxU4WU7G7BUw3dlJU0:-NHxU4WU7G7BUw3dlJU1:b-m1y96g) ([raw file](https://github.com/harness/harness-core/blob/c651ac0f00f298bfee81dadce67d9efd292d232a/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/step/IACMPMSStepPlanCreator.java#L18)):*
> ```Java
> 
> @OwnedBy(HarnessTeam.IACM)
> public class IACMPMSStepPlanCreator extends GenericStepPMSPlanCreator {
> ```

Dont include PMS keyword as its not pipeline-service step, you can ignore pms keyword
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java` line 30 at r11](https://reviewable.io/reviews/harness/harness-core/39942#-NHxUBYQ2_jhZ9eg63tW:-NHxUBYQ2_jhZ9eg63tX:b-9qm9kq) ([raw file](https://github.com/harness/harness-core/blob/c651ac0f00f298bfee81dadce67d9efd292d232a/310-iacm-manager/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java#L30)):*
> ```Java
>     engineSteps.put(InitializeTaskStep.STEP_TYPE, InitializeTaskStep.class);
>     engineSteps.put(IntegrationStageStepPMS.STEP_TYPE,
>         IntegrationStageStepPMS.class); // This seems to be the STAGE STEP per ser. So the stage is treated as a STEP
> ```

@harshjain12 ci stage in another service will be running?



_Code quote:_
```Java
IntegrationStageStepPMS.class); 
```

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 10 files at r8, 1 of 8 files at r12.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942)* status: 19 file(s) to review. 

___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java` line 114 at r3](https://reviewable.io/reviews/harness/harness-core/39942#-NGyrhIL7njJzVXZRq_8:-NI6kgIYE2tIy1CbKKoS:b-6evkk) ([raw file](https://github.com/harness/harness-core/blob/27f9656c2cf59c2eeaf0d1b55f768bd847578970/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/stage/IACMStagePMSPlanCreator.java#L114)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

To be check with CI Team cc: @harshjain12 
</blockquote></details>

done, ci team verified
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/step/IACMPMSStepPlanCreator.java` line 18 at r11](https://reviewable.io/reviews/harness/harness-core/39942#-NHxU4WU7G7BUw3dlJU0:-NI6kjiN0ROqPkZr9Eh1:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/c651ac0f00f298bfee81dadce67d9efd292d232a/310-iacm-manager/src/main/java/io/harness/iacm/plan/creator/step/IACMPMSStepPlanCreator.java#L18)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Dont include PMS keyword as its not pipeline-service step, you can ignore pms keyword
</blockquote></details>

done
___
*[`310-iacm-manager/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java` line 30 at r11](https://reviewable.io/reviews/harness/harness-core/39942#-NHxUBYQ2_jhZ9eg63tW:-NI6kksm7iMifyTtMH-H:b-7rd5hy) ([raw file](https://github.com/harness/harness-core/blob/c651ac0f00f298bfee81dadce67d9efd292d232a/310-iacm-manager/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java#L30)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@harshjain12 ci stage in another service will be running?
</blockquote></details>

done ci team verified

**Review by harshjain12 — APPROVED** [review]
:lgtm:

> Reviewed 12 of 16 files at r11, 8 of 8 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39942#-:-NI7Ddeh1gDlvLRcoMhz:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**